packages/select/src/components/select/dropdownRenderer.ts

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the terms of the LICENSE file distributed with this project.
+ */
+
+import * as React from "react";
+
+/**
+ * An object describing how to render the contents of a dropdown.
+ * A `dropdownRenderer` receives this object as its sole argument.
+ */
+export interface IDropdownRendererProps<T> {
+    /** Array of all items in the list. */
+    items: T[];
+
+    /**
+     * A ref handler that should be attached to the dropdown's outermost HTML element.
+     * This is required for the active item to scroll into view automatically.
+     */
+    itemsParentRef: (ref: HTMLElement) => void;
+
+    /**
+     * Call this function to render an item.
+     * This retrieves the modifiers for the item and delegates actual rendering
+     * to the owner component's `itemRenderer` prop.
+     */
+    renderItem: (item: T, index?: number) => JSX.Element | null;
+
+    /** React child to render when filtering items returns zero results. */
+    noResults?: React.ReactChild;
+
+    /** React child to render when query is empty. */
+    initialContent?: React.ReactChild;
```

Comment:
given that `noResults` and `initialContent` are optional in the first place and kinda specific to semantics of the default renderer implementation, it feels a little silly to pass them through here. while i'm not opposed to this, how would you feel about not supporting these two props through `menuRenderer` because that introduces a weird level of indirection:

```jsx
class Cmp extends React.Component {
    render() {
        return <Select menuRenderer={this.renderMenu} noResults={<NoResults />} ... />;
    }
   
    renderMenu(props) {
        // weird reference to a variable I defined 4 lines above in a separate method
        return <div>{props.noResults}</div>;
        // versus:
        return <div><NoResults /></div>;
    }
}
```

Reply:
I agree that this feels silly but the level of indirection is pretty much the same as with defining `itemRenderer` on the `<Select>` (which appears to be missing in your example) and then consuming it in `renderMenu` -- in my opinion that is the worse offender and the one we should be addressing, if we want to go that way. 

As far as I can tell the only way to get rid of the indirection altogether is to let `renderMenu` do its own thing, passing it only the bare essential props like the item array and possibly (as you suggested further up) the `query` itself. But then you have this strange situation where some props (here: `noResults`, `initialContent` and `itemRenderer`) are conditionally ignored. I'm not sure if this should be considered an anti-pattern.

Either way, I'm ok with leaving `noResults` and `initialContent` out of the menu renderer props if it makes things simpler.